drivers: gpio: gd32: add dependency on the EXTI

soburi opened the pull request:
The GPIO pin interruption depends on the EXTI.
Add "select" to clarify the dependency.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

cfriedt commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -46,7 +46,9 @@ struct gpio_gd32_config {
 	struct gpio_driver_config common;
 	uint32_t reg;
 	uint16_t clkid;
+#ifdef CONFIG_GD32_EXTI
```
It looks as though this will be padded out to 32-bits in any case. If you remove the `#ifdef` here, then you can switch to using `if (IS_ENABLED(CONFIG_GD32_EXTI)` elsewhere and remove the `#ifdef` around `static void gpio_gd32_isr(uint8_t line, void *arg)`

soburi replied:
> It looks as though `clkid` will be padded out to 32-bits if `CONFIG_GD32_EXTI=n`, so the `#ifdef` can be removed here.

Insightful!
Removed the `#ifdef`.



> Then you can switch to using `if (IS_ENABLED(CONFIG_GD32_EXTI)` elsewhere and remove the `#ifdef` around `static void gpio_gd32_isr(uint8_t line, void *arg)`

Commenting in follows topics.

cfriedt commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -286,6 +290,7 @@ static int gpio_gd32_pin_interrupt_configure(const struct device *port,
 					     enum gpio_int_mode mode,
 					     enum gpio_int_trig trig)
 {
+#ifdef CONFIG_GD32_EXTI
```
`if (IS_ENABLED(CONFIG_GD32_EXTI))`

soburi replied:
Fixed it.

cfriedt commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -350,8 +356,10 @@ static int gpio_gd32_init(const struct device *port)
 
 	(void)clock_control_on(GD32_CLOCK_CONTROLLER,
 			       (clock_control_subsys_t *)&config->clkid);
+#ifdef CONFIG_GD32_EXTI
```
`if (IS_ENABLED(CONFIG_GD32_EXTI))`

soburi replied:
Fixed it.

cfriedt commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -55,6 +57,7 @@ struct gpio_gd32_data {
 	sys_slist_t callbacks;
 };
 
+#ifdef CONFIG_GD32_EXTI
```
`#ifdef` can be removed

soburi replied:
This `#ifdef` suppressing `unused-function` warning.
It should be better not to remove it.

cfriedt commented on drivers/gpio/Kconfig.gd32:
```diff
@@ -5,5 +5,6 @@ config GPIO_GD32
 	bool "GD32 GPIO driver"
 	default y
 	depends on DT_HAS_GD_GD32_GPIO_ENABLED
+	imply GD32_EXTI
```
Should this be enabled by default?

soburi replied:
Yes, it is.
Interrupts handling needs this option.
The callback function of the GPIO API will not work If this option is disabled.
This PR is intended mainly to clarify dependencies.

cfriedt replied:
If it's required, should be `depends on`. `imply` means that it is optional but should be the default

soburi replied:
I added more change gpio_add_callback/gpio_remove_callback api returns -ENOTSUP if GD32_EXTI disabled.

It make allow to disable GD32_EXTI without API contradiction.

https://github.com/zephyrproject-rtos/zephyr/pull/50879/files#diff-606da3fe7a304f5544526be564097f77dc7db8e627c004fc62b8eed50e1b45dbR335-R341

henrikbrixandersen reviewed (dismissed):
LGTM

nandojve commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -55,6 +55,7 @@ struct gpio_gd32_data {
 	sys_slist_t callbacks;
 };
 
+#if (IS_ENABLED(CONFIG_GD32_EXTI))
```
I'm not sure but should it be as below?

```diff
-#if (IS_ENABLED(CONFIG_GD32_EXTI))
+if (IS_ENABLED(CONFIG_GD32_EXTI)) {
...
-#endif
+}
```

soburi replied:
It's a beautiful solution!

The if-else condition must determine at compile time.
I worry that the intention maybe not have been conveyed to the code reader, adding some comments.

gmarull reviewed (commented):
So before reviewing, why do we need this change? Why would a new GD platform need that?

gmarull reviewed (changes requested):
-1 if this is needed for https://github.com/zephyrproject-rtos/zephyr/pull/49886

soburi commented:
@gmarull ,

> -1 if this is needed for #49886

I found this problem in work about #49886, but This PR is not related to #49886 and is also not required by.

It is a problem that no dependency defines in Kconfig even though gd32_gpio.c depends on gd32_exti.c. 

The simple way to solve this is to use `select GD32_EXTI` as @cfriedt says.
But making it able to compile without GD32_EXTI is the preferable way, I think, for the following reasons.

- It does not violate API spec. (gpio_pin_interrupt_configure can return -ENOTSUP.)
- It makes providing more options to users. (User can disable EXTI if no use GPIO callbacks.)
- The default behavior has not changed. (`imply` make EXTI enable default.)

I agree there is no solid reason to merge this patch.
Because it is useful only in rare cases.
But I think it is good to fix it from the viewpoint of the integrity of dependency.

nandojve commented:
Hi @gmarull ,

Do you still with same position or @soburi justification is OK?
We need decide if we will keep it or not. I'm fine with justification since it is not about #49886.

gmarull commented on drivers/gpio/gpio_gd32.c:
```diff
@@ -286,43 +286,53 @@ static int gpio_gd32_pin_interrupt_configure(const struct device *port,
-		if (ret < 0) {
-			return ret;
-		}
+	/*
+	 * The if-else clause intends to eliminate dependency on EXTI.
+	 * For suppressing code generation of the else part at compile time,
+	 * the condition needs to determine statically.
+	 */
+	if (IS_ENABLED(CONFIG_GD32_EXTI)) {
+		if (mode == GPIO_INT_MODE_DISABLED) {
```
IMO, what should be done is the following:

1. Add a Kconfig option to build the GD32 driver w/o interrupt support, e.g. `CONFIG_GPIO_GD32_INT_ENABLE`
2. Such option should be enabled by default, and select `GD32_EXTI` (intc drivers do not have a class, so we need to select on a driver basis)
3. Just ifdefout code if `CONFIG_GPIO_GD32_INT_ENABLE=n`. I've personally never liked `IS_ENABLED` because it can't be universally used without getting compile errors in some cases.

But still, I feel it is wrong to add such option at a driver level. It should likely be a subsystem setting to build GPIO drivers with/without interrupt support.

soburi replied:
It is as you say. I look at the implementation of other GPIO drivers, and none of them enable or disable .pin_interrupt_configure in the settings. So I would like to withdraw this idea.
However, I think that it is better to clarify the dependency on EXTI,
I have changed the fix (and PR title).
Could you please confirm this?

soburi committed:
drivers: gpio: gd32: add dependency on the EXTI

The GPIO pin interruption depends on the EXTI.
Add "select" to clarify the dependency.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

cameled commented on drivers/gpio/Kconfig.gd32:
```diff
@@ -5,5 +5,6 @@ config GPIO_GD32
 	bool "GD32 GPIO driver"
 	default y
 	depends on DT_HAS_GD_GD32_GPIO_ENABLED
+	select GD32_EXTI
```
EXTI provide some basic capability for Zephyr, thread schedule, error handle, external event receive, etc. GPIO is an optional one.
There no tightly relation between them. So I think we can ignore the depends for these, EXTI will always accessable.

cameled commented:
Hi @nandojve, @gmarull, please help to confirm the new changes.
